bt-core/src/main/java/bt/torrent/messaging/DefaultMessageRouter.java

```diff
@@ -74,47 +66,47 @@ public void unregisterMessagingAgent(Object agent) {
 
     @SuppressWarnings("unchecked")
     private void addConsumers(List<MessageConsumer<?>> messageConsumers) {
+        if (messageConsumers.isEmpty()) {
+            return;
+        }
+
+        synchronized (this) {
+            List<MessageConsumer<Message>> newGenericConsumers = new ArrayList<>(this.genericConsumers);
+            Map<Class<?>, Collection<MessageConsumer<?>>> newTypedConsumers = new HashMap<>(this.typedConsumers);
 
-        List<MessageConsumer<Message>> genericConsumers = new ArrayList<>();
-        Map<Class<?>, Collection<MessageConsumer<?>>> typedMessageConsumers = new HashMap<>();
-
-        messageConsumers.forEach(consumer -> {
-            Class<?> consumedType = consumer.getConsumedType();
-            if (Message.class.equals(consumedType)) {
-                genericConsumers.add((MessageConsumer<Message>) consumer);
-            } else {
-                typedMessageConsumers.computeIfAbsent(consumedType, k -> new ArrayList<>()).add(consumer);
-            }
-        });
-
-        synchronized (changesLock) {
-            this.changes.add(() -> {
-                this.genericConsumers.addAll(genericConsumers);
-                typedMessageConsumers.keySet().forEach(key -> this.typedConsumers
-                        .computeIfAbsent(key, k -> new ArrayList<>()).addAll(typedMessageConsumers.get(key))
-                );
+            messageConsumers.forEach(consumer -> {
+                Class<?> consumedType = consumer.getConsumedType();
+                if (Message.class.equals(consumedType)) {
+                    newGenericConsumers.add((MessageConsumer<Message>) consumer);
+                } else {
+                    newTypedConsumers.computeIfAbsent(consumedType, k -> new ArrayList<>()).add(consumer);
+                }
             });
+
+            this.genericConsumers = newGenericConsumers;
+            this.typedConsumers = newTypedConsumers;
         }
     }
 
     private void addProducers(Collection<MessageProducer> producers) {
-        synchronized (changesLock) {
-            this.changes.add(() -> {
-                this.producers.addAll(producers);
-            });
+        if (producers.isEmpty()) {
+            return;
+        }
+
+        synchronized (this) {
+            List<MessageProducer> newProducers = new ArrayList<>(this.producers);
```

Comment:
To avoid double allocation:
```suggestion
            List<MessageProducer> newProducers = new ArrayList<>(this.producers.size() + producers.size() + 1);
```